Filter Expression support

martinsumner opened the pull request:
Allow indexes to be evaluated into a Key/Value map using an eval pipeline, and then filtered using a filter expression

martinsumner committed:
Extend perf_SUITE

This is v6 of the perf_SUITE tests.  The test adds a complex index entry to every object, and then adds a new test phase to test regex queries.

There are three profiles added so the full, mini and profiling versions of perf_SUITE can be run without having to edit the file itself:

e.g. ./rebar3 as perf_mini do ct --suite=test/end_to_end/perf_SUITE

When testing as `perf_prof` summarised versions of the eprof results are now printed to screen.

The volume of keys within the full test suite has been dropped ... just to make life easier so that test run times are not excessively increase by the new features.

martinsumner committed:
Switch to re2 library

The re2 library, though not offering the complete functionality of pcre, is known to be much faster in many scenarios.

Switch to re2 in preparation for the OTP switch to re2 in OTP28

martinsumner committed:
Load chunk in spawned processes

Assume to make the job of gs easier - name makes a massive difference to load time in OTP 24.

martinsumner committed:
Correctly account for pause

alos try and improve test stability by increasing pause

martinsumner committed:
Merge branch 'mas-d31-i433perfSUITE' into mas-d31-i433re2

martinsumner committed:
Support both re2 and pcre

martinsumner committed:
Add microstate accounting to profile

martinsumner committed:
Add memory tracking during test phases

Identify and log out memory usage by test phase

martinsumner committed:
Merge branch 'mas-d31-i433perfSUITE' into mas-d31-i433re2

ThomasArts committed:
Use macros instead (#437)

martinsumner committed:
Don't print memory to screen in standard ct test

martinsumner committed:
Merge branch 'mas-d31-i433perfSUITE' into mas-d31-i433re2

martinsumner committed:
Initial support for capture in regex

Allow for a capturing regular expression to be passed with a filter function that will filter based on the captured output.

Also allows a specific captured term to be returned as the returned term (rather than the whole index term)

martinsumner committed:
Create leveled_filter.erl

First draft of allowing an externally provided comparison expression to be validated, parsed and evaluated

martinsumner committed:
Make binary comparisons in leveled_filter

Required as re2 can only output binary not string in captures.

martinsumner committed:
Add Filter Expression support

Allow for capture with either delimited index terms, or with regex capture, and then filtering of captured attributes using a logical expression.

The xref check `locals_not_used` is now ignored due to issues with auto-generated module.

martinsumner committed:
Initial lexer/parser for eval pipeline

Allow for the generation of projected attributes from an index term

martinsumner committed:
Update and extend the eval expression

Allow for additional functions in eval.  Change the leveled query API to generalise application of eval/filter expressions.

martinsumner committed:
Extend testing

martinsumner committed:
Add regex eval function to pipeline

Remove `capture` option - as all can be achieved via eval/filter.

martinsumner committed:
Add support for combination queries on same snapshot point

Allow for return_keys of logical combinations of queries

ThomasArts committed:
First version of QuickCheck properties for setop lang and filter lang

martinsumner committed:
Remove re2

martinsumner committed:
Modify filter

The Filter should expect to compare BETWEEN low AND high ... supporting high AND low may be confusing.

The filter should allow for identifier to be in any location in a comparator i.e.

string BETWEEN identifier AND identifier

integer > identifier

This make check some validity constraints harder at parse time (i.e. that High >= Low), but makes tse result more compatible.

For ease, of implementation IN can only be used for a  comparison between a string and a list of strings (with the identifier representing either part).

ThomasArts committed:
Update filterlang eqc property

ThomasArts committed:
Updated quickcheck for filter language

now passes 10k tests

ThomasArts committed:
improvements to implementation

ThomasArts committed:
Also improve eval lexer

martinsumner committed:
Be more specific re integer types

Differentiate between positive/negative integers and non-negative integers.  Stop parsing of unusable split/index values.

martinsumner committed:
Basic unicode testing

Works on index terms of type unicode_binary()

martinsumner committed:
Expand unicode test

ThomasArts committed:
small shrink improvement for filterlang generator

ThomasArts committed:
fix lexer to get rid of '<>'

ThomasArts committed:
Add eqc property for evallang

ThomasArts committed:
runtime type errors in eval

ThomasArts commented on rebar.config:
```diff
@@ -4,11 +4,11 @@
 
 {cover_excl_mods,
-  [testutil,
+  [leveled_filterlexer, leveled_filterparser, leveled_evallexer, leveled_evalparser,
```
What is the reason to exclude parser and lexer from coverage?
When I ran quickcheck property, I cover all code apart from some error cases. Was adding needless negative tests the reason?

martinsumner replied:
It is the auto-generated files that were the problem.  The eunit/ct tests cover leveled_filter, and leveled_eval modules - but there was a large number of coverage holes in the auto-generated files.

Seen as they are auto-generated, I din't think they were relevant to coverage.

ThomasArts commented on src/leveled_codec.erl:
```diff
@@ -112,10 +112,17 @@
 -type maybe_lookup() ::
         lookup|no_lookup.
+-type actual_regex() ::
+        {re_pattern, term(), term(), term(), term()}.
```
would it not be nicer to use `re:mp()` here, to hide the details?

martinsumner replied:
The type is not exported until OTP 27 I think - so you can't refer to it.

ThomasArts replied:
in leveled_runner you replaced it by `any()`, but I prefer here to wait until OTP27 is lowest supported and replace it then.

ThomasArts commented on src/leveled_eval.erl:
```diff
@@ -0,0 +1,648 @@
+%%%============================================================================
+
+generate_eval_function(EvalString, Substitutions) ->
+    {ok, ParsedEval} = generate_eval_expression(EvalString, Substitutions),
```
```suggestion
    CodePointList = unicode:characters_to_list(EvalString),
    {ok, ParsedEval} = generate_eval_expression(CodePointList, Substitutions),
```

ThomasArts commented on src/leveled_eval.erl:
```diff
@@ -0,0 +1,648 @@
+    end.
+
+generate_eval_expression(EvalString, Substitutions) ->
+    {ok, Tokens, _EndLine} = leveled_evallexer:string(EvalString),
```
```suggestion
    CodePointList = unicode:characters_to_list(EvalString),
    {ok, Tokens, _EndLine} = leveled_evallexer:string(CodePointList),
```

ThomasArts commented on src/leveled_eval.erl:
```diff
@@ -0,0 +1,648 @@
+        " | map($birthday, <=, ((\"0430\", 2024)), 2023, $yoc)"
+        " | subtract($yoc, $yob, $age)"
+        " | add($age, 1, $age_next)"
+        " | to_string($age, $age)"
```
Didn't realize one can overload variables to have different types... probably not the nicest thing to use in an example.

ThomasArts commented on src/leveled_evallexer.xrl:
```diff
@@ -0,0 +1,55 @@
+%% Lexer for eval expressions
+
+Definitions.
+WhiteSpace  = ([\t\f\v\r\n\s]+)
```
When reading the documentation, I would say `\s` subsumes the others.

ThomasArts commented on src/leveled_evallexer.xrl:
```diff
@@ -0,0 +1,55 @@
+=       : {token, {comparator, '=', TokenLine}}.
+<       : {token, {comparator, '<', TokenLine}}.
+>       : {token, {comparator, '>', TokenLine}}.
+<>      : {token, {comparator, '<>', TokenLine}}.
```
```suggestion
```

martinsumner committed:
Update src/leveled_evallexer.xrl

Co-authored-by: Thomas Arts <thomas.arts@quviq.com>

martinsumner committed:
Update src/leveled_eval.erl

Co-authored-by: Thomas Arts <thomas.arts@quviq.com>

martinsumner committed:
Merge remote-tracking branch 'quviq/mas-d31-mas.i1433-filterexpression' into mas-d31-mas.i1433-filterexpression

martinsumner committed:
Revert "runtime type errors in eval"

This reverts commit 1b320c6686594c0577eab14731ccadd3bef93548.

martinsumner committed:
Non-empty strings and type handling

Update tot he lexers, to require strings to be non-empty, which rules out a raft of potential failure scenarios during evaluation.

Handle unexpected types in eval functions by not creating  the attribute rather than throwing an exception.

Add a new ends_with function, that behaves as begins_with in reverse.

martinsumner committed:
Setop parser to use only set operation names

Avoid confusion over meaning of NOT

ThomasArts committed:
Proposed changes to handle indexed sets

ThomasArts committed:
quickcheck properties for setop language

martinsumner committed:
Update setop to use maps as input

Don't rely on ordering of inputs, use a numeric identifier as a key

martinsumner committed:
Merge branch 'develop-3.4' into mas-d31-mas.i1433-filterexpression

martinsumner committed:
Remove duplications following merge

martinsumner committed:
Merge branch 'develop-3.4' into mas-d31-mas.i1433-filterexpression

martinsumner committed:
Additional changes to binary keys in test

martinsumner committed:
SubKeys must be binary()/null

SubKey of integer() used - falls foul of stronger checking

martinsumner committed:
Merge branch 'develop-3.4' into mas-d31-mas.i1433-filterexpression

martinsumner committed:
Merge branch 'develop-3.4' into mas-d31-mas.i1433-filterexpression

martinsumner committed:
Add error output when generating function from unparsable input

martinsumner committed:
Binary split not string split

Using the binary match/split to break up binary strings is a faster than using string split, as pre-compiled patterns can be used (saving 0.16 microsec per application).

compiled patterns can be passed across nodes, so pre-compiling them in the parser makes the function non-specific.  So instead, compile on demand.  If, as expected, this process is to check many times - at least have it compile only the once.

Examining the eprof profile also indicated significant time in lists:sublist - so the process of updating the map has been simplified.  Empty attributes will not now be added - and in the case of regex, attributes are only added if all are produced as expected.

martinsumner committed:
Fix types

martinsumner committed:
Merge branch 'develop-3.4' into mas-d31-mas.i1433-filterexpression
